### PR TITLE
Add icon and resource files for windows

### PR DESCRIPTION
**Pull request by ntzrmtthihu777**
Not yet integrated into the cmake build, will look into doing that soonish.
As of right now this works with your ad-hoc build script you mentioned in #35 
with a small bit of tweaking.

### CODE REVIEW

**Comment by Ancurio** [comment]
Hmm. I actually want to solve this differently, by having the previous icon loading code only run on Linux (we also want to avoid it on OSX because it clashes with the bundle-provided high resolution icons).  
I'll push a commit soon, then you can base your changes ontop of that.

File: src/main.cpp [hunk 1]
```diff
@@ -40,7 +40,11 @@
 
 #include "binding.h"
 
+#ifdef WIN32 || WIN64
+#include "resource.h"
```
Comment (Ancurio) [inline src/main.cpp]: I've worked with resource files in the past, but don't quite remember the details. Just looking at this, all that happens is a define being included. Why is this necessary?

File: assets/resource.h [hunk 2]
```diff
@@ -0,0 +1 @@
+#define IDI_APPICON 101
```
Comment (Ancurio) [inline assets/resource.h]: Having a define in a separate header that is used only once seems a bit overkill to me. The define itself is okay since it explains the magic value, I just don't get the extra header.

File: src/main.cpp [hunk 3]
```diff
@@ -40,7 +40,7 @@
 
 #include "binding.h"
 
-#ifdef WIN32 || WIN64
+#if defined(WIN32) && defined(WIN64)
```
Comment (Ancurio) [inline src/main.cpp]: I've used `__WINDOWS__` elsewhere in mkxp before. Could you check if that works for both architectures?

**Comment by Ancurio** [comment]
Also in the future, consider squashing fixup commits (and force-pushing to your branch if required); no need to have these in the official history.

**Comment by ntzrmtthihu777** [comment]
To be honest I have no idea about the squashing and such, really need to brush up on my git-fu. Reason I have a separate header is the potential of adding manifests and version information and whatnot, its just a starting point.

On another note, I've set up the ugliest `Makefile` ever, but it works and is much faster than the ad-hoc build script. But, this is mostly a stop-gap, as I'd like to get a windows build to work with cmake alone if possible.

**Comment by Ancurio** [comment]
Pushed 6349146e0198c995cadb923a7991576064df7e90

> On another note, I've set up the ugliest Makefile ever, but it works and is much faster than the ad-hoc build script. But, this is mostly a stop-gap, as I'd like to get a windows build to work with cmake alone if possible.

Yeah, my script recompiles everything everytime :) I used `ccache` internally to work around that, but a Makefile is of course the saner approach.

**Comment by ntzrmtthihu777** [comment]
For x86_64-w64-mingw32-g++
```c
#define _WIN32 1
#define _WIN64 1
#define __WINT_MAX__ 0xffff
#define __WINT_MIN__ 0
#define __WIN32 1
#define __WIN64 1
#define __WINNT 1
#define __WINNT__ 1
#define __WIN32__ 1
#define __SIZEOF_WINT_T__ 2
#define WIN32 1
#define WIN64 1
#define __WINT_TYPE__ short unsigned int
#define WINNT 1
#define __WIN64__ 1
```
and i686-w64-mingw32-g++
```c
#define _WIN32 1
#define __WINT_MAX__ 0xffff
#define __WINT_MIN__ 0
#define __WIN32 1
#define __WINNT 1
#define __WINNT__ 1
#define __WIN32__ 1
#define __SIZEOF_WINT_T__ 2
#define WIN32 1
#define __WINT_TYPE__ short unsigned int
#define WINNT 1
```
the diff:
```diff
--- i686.h	2017-03-04 10:14:53.610710212 +0000
+++ x86_64.h	2017-03-04 10:14:42.307376535 +0000
@@ -1,11 +1,15 @@
 #define _WIN32 1
+#define _WIN64 1
 #define __WINT_MAX__ 0xffff
 #define __WINT_MIN__ 0
 #define __WIN32 1
+#define __WIN64 1
 #define __WINNT 1
 #define __WINNT__ 1
 #define __WIN32__ 1
 #define __SIZEOF_WINT_T__ 2
 #define WIN32 1
+#define WIN64 1
 #define __WINT_TYPE__ short unsigned int
 #define WINNT 1
+#define __WIN64__ 1
```

**Comment by Ancurio** [comment]
https://hg.libsdl.org/SDL/rev/cb37776926ad implies that `__WINDOWS__` works on both.

**Comment by ntzrmtthihu777** [comment]
There we go. Well, is `__WINDOWS__` supplied by SDL or by the compiler? Suppose that works, since `SDL.h` is included on the very file in question.

File: src/main.cpp [hunk 4]
```diff
@@ -40,7 +40,11 @@
 
 #include "binding.h"
 
+#ifdef(__WINDOWS__)
+#include "resource.h"
+#else
 #include "icon.png.xxd"
+#endif
```
Comment (Ancurio) [inline src/main.cpp]: No need for the else I think? Or do you get compiler warnings if icon.png is included on windows?

File: src/main.cpp [hunk 5]
```diff
@@ -166,6 +170,24 @@ static void showInitError(const std::string &msg)
...
+
+	SDL_Surface *iconImg = IMG_Load_RW(iconSrc, SDL_TRUE);
+
+	if (iconImg)
+	{
+		SDL_SetWindowIcon(win, iconImg);
+		SDL_FreeSurface(iconImg);
+	}
+}
+
```
Comment (Ancurio) [inline src/main.cpp]: Don't understand why my commit shows up here.

File: src/main.cpp (hunk 4 above)
Comment (ntzrmtthihu777) [inline src/main.cpp]: No, no warnings, but it's a waste of space in the binary imho.

File: src/main.cpp (hunk 5 above)
Comment (ntzrmtthihu777) [inline src/main.cpp]: Honestly I probably butchered the process of squashing.

File: src/main.cpp (hunk 4 above)
Comment (Ancurio) [inline src/main.cpp]: Then change the include to only happen on Linux, so it's consistent (it's just 1kb, but I don't care either way).

Comment (ntzrmtthihu777) [inline src/main.cpp]: K, done. It may only be 1kb now, but someone may build their own with a different icon of higher quality/size, so it could blow up a bit. Better safe than sorry.

File: assets/resource.h (hunk 2 above)
Comment (ntzrmtthihu777) [inline assets/resource.h]: Its mostly like this because its how I've seen it done, and it may only be one line atm, but you can add a lot more information into a windows executable here.

File: src/main.cpp [hunk 6]
```diff
@@ -40,7 +40,11 @@
 
 #include "binding.h"
 
+#ifdef __WINDOWS__
+#include "resource.h"
+#elseif __LINUX__
```
Comment (carstene1ns) [inline src/main.cpp]: I think it is better to use `#elif` here. It is shorter, more common and consistent with the rest of the mkxp source code. (Also, I did not find documentation if `#elseif` is even supported under MSVC, but that is just an assumption)

**Comment by ntzrmtthihu777** [comment]
@carstene1ns

Suppose your right. To be frank I don't care about what MSVC supports at
all, but I agree #elif is nicer looking and shorter.

File: src/main.cpp (hunk 6 above)
Comment (Ancurio) [inline src/main.cpp]: @carstene1ns That would be because it [doesn't exist](http://stackoverflow.com/questions/36137376/elseif-vs-elif-c-c-preprocessor) :)

Comment (carstene1ns) [inline src/main.cpp]: Yeah, I was not entirely sure, so did a quick google search and found some entries. Therefore I assumed GCC supports it.

**Comment by ntzrmtthihu777** [comment]
@Ancurio well?

**Comment by Ancurio** [comment]
Sorry, didn't see that the commit changed. LGTM.  
What image is inside the .ico? Is it the same as `icon.png`?

**Comment by ntzrmtthihu777** [comment]
@Ancurio yeah, same as ico.png
